lib/graph_gateway.go

```diff
@@ -0,0 +1,482 @@
+package lib
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/filecoin-saturn/caboose"
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	leveldb "github.com/ipfs/go-ds-leveldb"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	exchange "github.com/ipfs/go-ipfs-exchange-interface"
+	"github.com/ipfs/go-libipfs/blocks"
+	"github.com/ipfs/go-libipfs/files"
+	"github.com/ipfs/go-libipfs/gateway"
+	"github.com/ipfs/go-namesys"
+	"github.com/ipfs/go-namesys/resolve"
+	ipfspath "github.com/ipfs/go-path"
+	nsopts "github.com/ipfs/interface-go-ipfs-core/options/namesys"
+	ifacepath "github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/ipld/go-car"
+	routinghelpers "github.com/libp2p/go-libp2p-routing-helpers"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+	"go.uber.org/multierr"
+	"io"
+	"net/http"
+)
+
+// type DataCallback = func(resource string, reader io.Reader) error
+// TODO: Don't use a caboose type, perhaps ask them to use a type alias instead of a type
+type DataCallback = caboose.DataCallback
+
+type CarFetcher interface {
+	Fetch(ctx context.Context, path string, cb DataCallback) error
+}
+
+type gwOptions struct {
+	ns namesys.NameSystem
+	vs routing.ValueStore
+	bs blockstore.Blockstore
+}
+
+// WithNameSystem sets the name system to use for the gateway. If not set it will use a default DNSLink resolver
+// along with any configured ValueStore
+func WithNameSystem(ns namesys.NameSystem) GraphGatewayOption {
+	return func(opts *gwOptions) error {
+		opts.ns = ns
+		return nil
+	}
+}
+
+// WithValueStore sets the ValueStore to use for the gateway
+func WithValueStore(vs routing.ValueStore) GraphGatewayOption {
+	return func(opts *gwOptions) error {
+		opts.vs = vs
+		return nil
+	}
+}
+
+// WithBlockstore sets the Blockstore to use for the gateway
+func WithBlockstore(bs blockstore.Blockstore) GraphGatewayOption {
+	return func(opts *gwOptions) error {
+		opts.bs = bs
+		return nil
+	}
+}
+
+type GraphGatewayOption func(gwOptions *gwOptions) error
+
+type GraphGateway struct {
+	fetcher      CarFetcher
+	blockFetcher exchange.Fetcher
+	routing      routing.ValueStore
+	namesys      namesys.NameSystem
+	bstore       blockstore.Blockstore
+	bsrv         blockservice.BlockService
+}
+
+func NewGraphGatewayBackend(f CarFetcher, blockFetcher exchange.Fetcher, opts ...GraphGatewayOption) (*GraphGateway, error) {
```

Comment:
Using this as a thread anchor for any misc questions, but overall this code is very much WIP (with a bunch of copy-paste) to try and get this all to work reasonably.

There's a lot of cleaning up and improvements to do here. Here are a few, but feel free to recommend more.

1. Incorrect internal abstraction usage
   - Caboose is using a blockstore, which means we have a proxy blockstore, but these are both the wrong abstraction. At best they are exchange.Fetcher's. This starts bleeding through the codebase here. Let's wrap them up nicely and isolate some of the crazy.
2. Copy pasted code from the BlocksGateway
  - We should separate out some of the blocks gateway components (e.g. the mutability pieces have nothing to do with blocks and should be separately usable)
   
Some todo item's to consider in other refactors that were noticed here:
    - go-path's reliance on go-fetcher to do sessions is .... bad. It effectively forces you to break up the session into "path" + other.
    - go-fetcher's requirement on using a blockservice is overkill, I basically had to copy-paste it so I could give it a NodeGetter, which is fine for now. Although even that should probably be an exchange.Fetcher rather than a NodeGetter